src/jdk.compiler/share/classes/com/sun/tools/javac/comp/Flow.java

```diff
@@ -772,72 +775,94 @@ private Set<Symbol> coveredSymbols(DiagnosticPosition pos, Type targetType,
                         if (label.isExpression()) {
                             JCExpression expr = (JCExpression) label;
                             if (expr.hasTag(IDENT) && ((JCIdent) expr).sym.isEnum())
-                                constants.add(((JCIdent) expr).sym);
+                                coveredSymbols.add(((JCIdent) expr).sym);
                         } else {
                             throw new AssertionError(label.getTag());
                         }
                     }
                 }
             }
-            for (Entry<Symbol, List<JCDeconstructionPattern>> e : categorizedDeconstructionPatterns.entrySet()) {
-                if (coversDeconstructionStartingFromComponent(pos, targetType, e.getValue(), 0)) {
-                    constants.add(e.getKey());
+            for (Entry<Symbol, List<JCRecordPattern>> e : deconstructionPatternsBySymbol.entrySet()) {
+                if (coversDeconstructionFromComponent(pos, targetType, e.getValue(), 0)) {
+                    coveredSymbols.add(e.getKey());
                 }
             }
-            return constants;
+            return coveredSymbols;
         }
 
-        private boolean coversDeconstructionStartingFromComponent(DiagnosticPosition pos,
-                                                                  Type targetType,
-                                                                  List<JCDeconstructionPattern> patterns,
-                                                                  int component) {
-            List<? extends RecordComponent> components = patterns.head.record.getRecordComponents();
+        private boolean coversDeconstructionFromComponent(DiagnosticPosition pos,
+                                                          Type targetType,
+                                                          List<JCRecordPattern> deconstructionPatterns,
+                                                          int component) {
+            //Given a set of record patterns for the same record, and a starting component,
+            //this method checks, whether the nested patterns for the components are exhaustive,
+            //i.e. represent all possible combinations.
+            //This is done by categorizing the patterns based on the type covered by the given
+            //starting component.
```

Comment:
Example needed here. For instance (I discussed this with @biboudis):

```
record Outer(R r) { };
sealed interface I { };
class A implements I { };
class B implements I { };
sealed interface R { };
record Foo(I i) implements R { }
record Bar(I i) implements R { }

switch (o) {
   case Outer(Foo(A), Foo(A)):
   case Outer(Foo(B), Foo(B)):
   case Outer(Foo(A), Foo(B)):
   case Outer(Foo(B), Foo(A)):
   case Outer(Bar(A), Bar(A)):
   case Outer(Bar(B), Bar(B)):
   case Outer(Bar(A), Bar(B)):
   case Outer(Bar(B), Bar(A)):
}
```

Which generates two sets:

```
   case Outer(Foo(A), Foo(A)):
   case Outer(Foo(B), Foo(B)):
   case Outer(Foo(A), Foo(B)):
   case Outer(Foo(B), Foo(A)):
```

And

```
   case Outer(Bar(A), Bar(A)):
   case Outer(Bar(B), Bar(B)):
   case Outer(Bar(A), Bar(B)):
   case Outer(Bar(B), Bar(A)):
```

Sorry for being pedantic - this code is tricky and I'm worried we'll all forget exactly how it works in 2 months :-)

Reply:
Sure. Will try to improve.